644 maintenance update page fixpack list

steven-chiu-bigstack opened the pull request:
#### What type of PR is this?

Feat

#### What this PR does / why we need it

Implement the Maintenance Update page fixpacks list.

#### Which issue(s) this PR fixes

Fixes #644

#### Special notes for your reviewer

The upload, install, rollback, and removal of fixpacks will be handled in other issues.

https://github.com/user-attachments/assets/97c4c8b2-5b0c-4f4c-a2e7-a5fad84737a1

#### Additional documentation

None

steven-chiu-bigstack committed:
chore: update api version

Signed-off-by: steven-chiu-bigstack <steven.chiu@bigstack.co>

steven-chiu-bigstack committed:
feat(web-app): maintenance update page fixpack list

Signed-off-by: steven-chiu-bigstack <steven.chiu@bigstack.co>

raven-pan commented on packages/cube-frontend-web-app/src/api/cosApi.ts:
```diff
@@ -5,6 +5,7 @@ import {
   DataCentersApi,
   EventsApi,
   FirmwaresApi,
```
Discussion: We can add `firmwaresApi` to the `cSpell.words` section of `.vscode/settings.json` to prevent VSCode typo hint.

steven-chiu-bigstack replied:
👌

raven-pan commented on packages/cube-frontend-web-app/src/pages/maintenance/update/fixpack/computeFixpacksActionState.test.ts:
```diff
@@ -0,0 +1,396 @@
+
+    beforeEach(() => {
+      const states = computeFixpacksActionState([
+        createFixpack({ current: StatusEnum.Available, isRollbackable: true }),
```
We could add `rollbackable` to our spellcheck dictionary.

<img width="613" height="246" alt="Screenshot 2025-09-02 at 3 51 30 PM" src="https://github.com/user-attachments/assets/a2240b66-4651-43c5-93b1-5511a2b705e3" />

steven-chiu-bigstack replied:
👌

raven-pan commented on packages/cube-frontend-web-app/src/pages/maintenance/update/fixpack/listFixpacksUtils.ts:
```diff
@@ -0,0 +1,52 @@
+export const searchParamsToQuery = (
+  searchParams: URLSearchParams,
+): ListFixpacksQuery => {
+  const currentPage = searchParams.get('pageNum')
```
This should be `page`, and we could refactor it to `searchParams.get(FixpackParamKeyEnum.CurrentPage)`

steven-chiu-bigstack replied:
You're absolute right, good catch! 🙏

raven-pan commented on packages/cube-frontend-web-app/src/pages/maintenance/update/fixpack/MaintenanceUpdateFixpackPage.tsx:
```diff
@@ -1,16 +1,145 @@
+                  const actionState = fixpacksActionStates[index]
+                  return (
+                    !!actionState &&
+                    renderAction(row, fixpacksActionStates[index])
```
We could just write as `renderAction(row, actionState)`

steven-chiu-bigstack replied:
Right, I missed this one. Thanks!🙏

raven-pan commented on packages/cube-frontend-web-app/src/pages/maintenance/update/fixpack/useListFixpacksQuery.ts:
```diff
@@ -0,0 +1,40 @@
+  const onItemsPerPageChange = (itemsPerPage: ItemsPerPage): void => {
+    setQuery((prev) => ({
+      ...prev,
+      pageSize: itemsPerPage,
```
Should reset the `pageNum` to 1 here.

steven-chiu-bigstack replied:
Good catch! I should've been more careful about the pagination logic. 🙏

raven-pan reviewed (approved):
Good work, I really like the way you test the `actionStates`.
Using emojis is much more readable and easy to find specific test cases.

I want to discussion:  It will be great  if we add `When a rollbackable fixpack in the middle is installing` and `When a rollbackable fixpack in the middle is rolling back` test cases for more comprehensive coverage. what do you think?

I just left some comments and discussion, feel free to merge this PR once they're fixed.
Thanks!

steven-chiu-bigstack committed:
fix(web-app): list fixpacks and list firmwares issues

Signed-off-by: steven-chiu-bigstack <steven.chiu@bigstack.co>

steven-chiu-bigstack commented:
> Good work, I really like the way you test the `actionStates`. Using emojis is much more readable and easy to find specific test cases.
> 
> I want to discussion: It will be great if we add `When a rollbackable fixpack in the middle is installing` and `When a rollbackable fixpack in the middle is rolling back` test cases for more comprehensive coverage. what do you think?
> 
> I just left some comments and discussion, feel free to merge this PR once they're fixed. Thanks!

Thanks for the great suggestions! In addition to the two test suites you mentioned, I'll also add test suites to block install/rollback when Ceph is unhealthy.

steven-chiu-bigstack committed:
feat(web-app): blocks fixpack install/rollback when ceph is unhealthy

Signed-off-by: steven-chiu-bigstack <steven.chiu@bigstack.co>

steven-chiu-bigstack commented:
Hi @raven-pan 

Thanks for the review! I've updated the code based on your comments and added the test suites I mentioned earlier. Let me know if you notice any issues. 🙏

raven-pan commented:
> Hi @raven-pan
> 
> Thanks for the review! I've updated the code based on your comments and added the test suites I mentioned earlier. Let me know if you notice any issues. 🙏

Thanks! Feel free to merge.

steven-chiu-bigstack committed:
feat(web-app): hides older fixpacks action buttons when newer unrollbackable fixpack is installed

Signed-off-by: steven-chiu-bigstack <steven.chiu@bigstack.co>

steven-chiu-bigstack committed:
refactor(web-app): maintenance update page code adjustment for the new cos general panel props

Signed-off-by: steven-chiu-bigstack <steven.chiu@bigstack.co>
